### DISCUSSION TITLE
Issue with pika.connection.update_secret

### DISCUSSION THREAD

**Discussion by rshandilya01** [pika/pika#1454] [Q&A] [answered]
I have an AWS Lambda rotator to rotate secret on RabbitMQ server running on Amazon MQ. Rotator has multiple steps including set and test secrets. I am using pika connection method - update_secret - to update the connection. This stage goes fine with no errors thrown. When testing secret with the new password, it throws an error. I am wondering if update_password is working correctly. I am rotating the password for the admin user. 

set_secret code - 
try:
    conn.update_secret(pending_dict['password'], 'set_secret AWS Secret Rotation by Lambda')
except pika.exceptions.ConnectionWrongStateError as connectionerror:
    logger.error(
        "set_secret: ConnectionWrongStateError while updating Secret on RabbitMQ Broker: Connection is not open or available",
        connectionerror)
except Exception as error:
    logger.error("set_secret: Error when updating Secret on RabbitMQ Broker:", error)
finally:
    conn.close()
logger.info("set_secret: Successfully set secret for %s." % arn)
 
Log statements - 
2023-11-16T11:39:50.841-06:00

Copy
[INFO]	2023-11-16T17:39:50.841Z	8fda8aeb-3351-4569-b44b-2fc419603c2d	Closing transport socket and unlinking: state=3; <ssl.SSLSocket fd=11, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6, laddr=('169.254.76.1', 46510), raddr=('10.191.69.188', 5671)>
[INFO] 2023-11-16T17:39:50.841Z 8fda8aeb-3351-4569-b44b-2fc419603c2d Closing transport socket and unlinking: state=3; <ssl.SSLSocket fd=11, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6, laddr=('169.254.76.1', 46510), raddr=('10.191.69.188', 5671)>

2023-11-16T11:39:50.842-06:00

Copy
[ERROR]	2023-11-16T17:39:50.842Z	8fda8aeb-3351-4569-b44b-2fc419603c2d	Connection workflow failed: AMQPConnectionWorkflowFailed: 1 exceptions in all; last exception - AMQPConnectorAMQPHandshakeError: ProbableAuthenticationError: Client was disconnected at a connection stage indicating a probable authentication error: ("ConnectionClosedByBroker: (403) 'ACCESS_REFUSED - Login was refused using authentication mechanism PLAIN. For details see the broker logfile.'",); first exception - None
[ERROR] 2023-11-16T17:39:50.842Z 8fda8aeb-3351-4569-b44b-2fc419603c2d Connection workflow failed: AMQPConnectionWorkflowFailed: 1 exceptions in all; last exception - AMQPConnectorAMQPHandshakeError: ProbableAuthenticationError: Client was disconnected at a connection stage indicating a probable authentication error: ("ConnectionClosedByBroker: (403) 'ACCESS_REFUSED - Login was refused using authentication mechanism PLAIN. For details see the broker logfile.'",); first exception - None

2023-11-16T11:39:50.859-06:00

Copy
[ERROR]	2023-11-16T17:39:50.842Z	8fda8aeb-3351-4569-b44b-2fc419603c2d	Error in _create_connection().
Traceback (most recent call last):
  File "/opt/python/pika/adapters/blocking_connection.py", line 451, in _create_connection
    raise self._reap_last_connection_workflow_error(error)
pika.exceptions.ProbableAuthenticationError: ConnectionClosedByBroker: (403) 'ACCESS_REFUSED - Login was refused using authentication mechanism PLAIN. For details see the broker logfile.'
[ERROR] 2023-11-16T17:39:50.842Z 8fda8aeb-3351-4569-b44b-2fc419603c2d Error in _create_connection(). Traceback (most recent call last): File "/opt/python/pika/adapters/blocking_connection.py", line 451, in _create_connection raise self._reap_last_connection_workflow_error(error) pika.exceptions.ProbableAuthenticationError: ConnectionClosedByBroker: (403) 'ACCESS_REFUSED - Login was refused using authentication mechanism PLAIN. For details see the broker logfile.'

2023-11-16T11:39:50.859-06:00

Copy
Traceback (most recent call last):
Traceback (most recent call last):

2023-11-16T11:39:50.859-06:00

Copy
File "/var/task/lambda_function.py", line 294, in get_mq_connection
File "/var/task/lambda_function.py", line 294, in get_mq_connection

2023-11-16T11:39:50.859-06:00

Copy
connection = pika.BlockingConnection(parameters)
connection = pika.BlockingConnection(parameters)

2023-11-16T11:39:50.859-06:00

Copy
File "/opt/python/pika/adapters/blocking_connection.py", line 360, in __init__
File "/opt/python/pika/adapters/blocking_connection.py", line 360, in __init__

2023-11-16T11:39:50.859-06:00

Copy
self._impl = self._create_connection(parameters, _impl_class)
self._impl = self._create_connection(parameters, _impl_class)

2023-11-16T11:39:50.859-06:00

Copy
File "/opt/python/pika/adapters/blocking_connection.py", line 451, in _create_connection
File "/opt/python/pika/adapters/blocking_connection.py", line 451, in _create_connection

2023-11-16T11:39:50.859-06:00

Copy
raise self._reap_last_connection_workflow_error(error)
raise self._reap_last_connection_workflow_error(error)

2023-11-16T11:39:50.859-06:00

Copy
pika.exceptions.ProbableAuthenticationError: ConnectionClosedByBroker: (403) 'ACCESS_REFUSED - Login was refused using authentication mechanism PLAIN. For details see the broker logfile.'
pika.exceptions.ProbableAuthenticationError: ConnectionClosedByBroker: (403) 'ACCESS_REFUSED - Login was refused using authentication mechanism PLAIN. For details see the broker logfile.'

**Comment by lukebakken** [comment]
What versions of the following software are you using:
* RabbitMQ and Erlang
* Python
* Pika

> When testing secret with the new password, it throws an error

Are you certain the new secret is available to RabbitMQ?

What does RabbitMQ log at the same time your application logs the `ProbableAuthenticationError` error?

**Reply by rshandilya01** [reply]
RabbitMQ Engine version = 3.11.20
Python = 3.7.3
Pika = 1.1.0

Yes - I had printed out the new secret in logs to confirm that it is available to RabbitMQ. 

Below is the log snippet leading upto the probable authentication error - 

<details>

```
2023-11-16T11:39:50.722-06:00

Copy
[INFO]	2023-11-16T17:39:50.722Z	8fda8aeb-3351-4569-b44b-2fc419603c2d	Socket connected: <socket.socket fd=11, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6, laddr=('169.254.76.1', 46510), raddr=('10.191.69.188', 5671)>
[INFO] 2023-11-16T17:39:50.722Z 8fda8aeb-3351-4569-b44b-2fc419603c2d Socket connected: <socket.socket fd=11, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6, laddr=('169.254.76.1', 46510), raddr=('10.191.69.188', 5671)>

2023-11-16T11:39:50.818-06:00

Copy
[INFO]	2023-11-16T17:39:50.818Z	8fda8aeb-3351-4569-b44b-2fc419603c2d	SSL handshake completed successfully: <ssl.SSLSocket fd=11, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6, laddr=('169.254.76.1', 46510), raddr=('10.191.69.188', 5671)>
[INFO] 2023-11-16T17:39:50.818Z 8fda8aeb-3351-4569-b44b-2fc419603c2d SSL handshake completed successfully: <ssl.SSLSocket fd=11, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6, laddr=('169.254.76.1', 46510), raddr=('10.191.69.188', 5671)>

2023-11-16T11:39:50.819-06:00

Copy
[INFO]	2023-11-16T17:39:50.819Z	8fda8aeb-3351-4569-b44b-2fc419603c2d	Streaming transport linked up: (<pika.adapters.utils.io_services_utils._AsyncSSLTransport object at 0x7f634e401280>, _StreamingProtocolShim: <SelectConnection PROTOCOL transport=<pika.adapters.utils.io_services_utils._AsyncSSLTransport object at 0x7f634e401280> params=<ConnectionParameters host=b-b3c1ac37-6053-4168-8d04-a3d6d87ccfef.mq.us-east-1.amazonaws.com port=5671 virtual_host=/ ssl=True>>).
[INFO] 2023-11-16T17:39:50.819Z 8fda8aeb-3351-4569-b44b-2fc419603c2d Streaming transport linked up: (<pika.adapters.utils.io_services_utils._AsyncSSLTransport object at 0x7f634e401280>, _StreamingProtocolShim: <SelectConnection PROTOCOL transport=<pika.adapters.utils.io_services_utils._AsyncSSLTransport object at 0x7f634e401280> params=<ConnectionParameters host=b-b3c1ac37-6053-4168-8d04-a3d6d87ccfef.mq.us-east-1.amazonaws.com port=5671 virtual_host=/ ssl=True>>).

2023-11-16T11:39:50.840-06:00

Copy
[INFO]	2023-11-16T17:39:50.840Z	8fda8aeb-3351-4569-b44b-2fc419603c2d	Aborting transport connection: state=1; <ssl.SSLSocket fd=11, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6, laddr=('169.254.76.1', 46510), raddr=('10.191.69.188', 5671)>
[INFO] 2023-11-16T17:39:50.840Z 8fda8aeb-3351-4569-b44b-2fc419603c2d Aborting transport connection: state=1; <ssl.SSLSocket fd=11, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6, laddr=('169.254.76.1', 46510), raddr=('10.191.69.188', 5671)>

2023-11-16T11:39:50.840-06:00

Copy
[INFO]	2023-11-16T17:39:50.840Z	8fda8aeb-3351-4569-b44b-2fc419603c2d	_AsyncTransportBase._initate_abort(): Initiating abrupt asynchronous transport shutdown: state=1; error=None; <ssl.SSLSocket fd=11, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6, laddr=('169.254.76.1', 46510), raddr=('10.191.69.188', 5671)>
[INFO] 2023-11-16T17:39:50.840Z 8fda8aeb-3351-4569-b44b-2fc419603c2d _AsyncTransportBase._initate_abort(): Initiating abrupt asynchronous transport shutdown: state=1; error=None; <ssl.SSLSocket fd=11, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6, laddr=('169.254.76.1', 46510), raddr=('10.191.69.188', 5671)>

2023-11-16T11:39:50.841-06:00

Copy
[INFO]	2023-11-16T17:39:50.840Z	8fda8aeb-3351-4569-b44b-2fc419603c2d	Deactivating transport: state=1; <ssl.SSLSocket fd=11, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6, laddr=('169.254.76.1', 46510), raddr=('10.191.69.188', 5671)>
[INFO] 2023-11-16T17:39:50.840Z 8fda8aeb-3351-4569-b44b-2fc419603c2d Deactivating transport: state=1; <ssl.SSLSocket fd=11, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6, laddr=('169.254.76.1', 46510), raddr=('10.191.69.188', 5671)>

2023-11-16T11:39:50.841-06:00

Copy
[INFO]	2023-11-16T17:39:50.841Z	8fda8aeb-3351-4569-b44b-2fc419603c2d	AMQP stack terminated, failed to connect, or aborted: opened=False, error-arg=None; pending-error=ConnectionClosedByBroker: (403) 'ACCESS_REFUSED - Login was refused using authentication mechanism PLAIN. For details see the broker logfile.'
[INFO] 2023-11-16T17:39:50.841Z 8fda8aeb-3351-4569-b44b-2fc419603c2d AMQP stack terminated, failed to connect, or aborted: opened=False, error-arg=None; pending-error=ConnectionClosedByBroker: (403) 'ACCESS_REFUSED - Login was refused using authentication mechanism PLAIN. For details see the broker logfile.'

2023-11-16T11:39:50.841-06:00

Copy
[ERROR]	2023-11-16T17:39:50.841Z	8fda8aeb-3351-4569-b44b-2fc419603c2d	Connection closed while authenticating indicating a probable authentication error
[ERROR] 2023-11-16T17:39:50.841Z 8fda8aeb-3351-4569-b44b-2fc419603c2d Connection closed while authenticating indicating a probable authentication error

2023-11-16T11:39:50.841-06:00

Copy
[INFO]	2023-11-16T17:39:50.841Z	8fda8aeb-3351-4569-b44b-2fc419603c2d	Connection setup terminated due to ProbableAuthenticationError: Client was disconnected at a connection stage indicating a probable authentication error: ("ConnectionClosedByBroker: (403) 'ACCESS_REFUSED - Login was refused using authentication mechanism PLAIN. For details see the broker logfile.'",)
[INFO] 2023-11-16T17:39:50.841Z 8fda8aeb-3351-4569-b44b-2fc419603c2d Connection setup terminated due to ProbableAuthenticationError: Client was disconnected at a connection stage indicating a probable authentication error: ("ConnectionClosedByBroker: (403) 'ACCESS_REFUSED - Login was refused using authentication mechanism PLAIN. For details see the broker logfile.'",)

2023-11-16T11:39:50.841-06:00

Copy
[ERROR]	2023-11-16T17:39:50.841Z	8fda8aeb-3351-4569-b44b-2fc419603c2d	AMQPConnector - reporting failure: AMQPConnectorAMQPHandshakeError: ProbableAuthenticationError: Client was disconnected at a connection stage indicating a probable authentication error: ("ConnectionClosedByBroker: (403) 'ACCESS_REFUSED - Login was refused using authentication mechanism PLAIN. For details see the broker logfile.'",)
[ERROR] 2023-11-16T17:39:50.841Z 8fda8aeb-3351-4569-b44b-2fc419603c2d AMQPConnector - reporting failure: AMQPConnectorAMQPHandshakeError: ProbableAuthenticationError: Client was disconnected at a connection stage indicating a probable authentication error: ("ConnectionClosedByBroker: (403) 'ACCESS_REFUSED - Login was refused using authentication mechanism PLAIN. For details see the broker logfile.'",)

2023-11-16T11:39:50.841-06:00

Copy
[ERROR]	2023-11-16T17:39:50.841Z	8fda8aeb-3351-4569-b44b-2fc419603c2d	AMQP connection workflow failed: AMQPConnectionWorkflowFailed: 1 exceptions in all; last exception - AMQPConnectorAMQPHandshakeError: ProbableAuthenticationError: Client was disconnected at a connection stage indicating a probable authentication error: ("ConnectionClosedByBroker: (403) 'ACCESS_REFUSED - Login was refused using authentication mechanism PLAIN. For details see the broker logfile.'",); first exception - None.
[ERROR] 2023-11-16T17:39:50.841Z 8fda8aeb-3351-4569-b44b-2fc419603c2d AMQP connection workflow failed: AMQPConnectionWorkflowFailed: 1 exceptions in all; last exception - AMQPConnectorAMQPHandshakeError: ProbableAuthenticationError: Client was disconnected at a connection stage indicating a probable authentication error: ("ConnectionClosedByBroker: (403) 'ACCESS_REFUSED - Login was refused using authentication mechanism PLAIN. For details see the broker logfile.'",); first exception - None.

2023-11-16T11:39:50.841-06:00

Copy
[ERROR]	2023-11-16T17:39:50.841Z	8fda8aeb-3351-4569-b44b-2fc419603c2d	AMQPConnectionWorkflow - reporting failure: AMQPConnectionWorkflowFailed: 1 exceptions in all; last exception - AMQPConnectorAMQPHandshakeError: ProbableAuthenticationError: Client was disconnected at a connection stage indicating a probable authentication error: ("ConnectionClosedByBroker: (403) 'ACCESS_REFUSED - Login was refused using authentication mechanism PLAIN. For details see the broker logfile.'",); first exception - None
[ERROR] 2023-11-16T17:39:50.841Z 8fda8aeb-3351-4569-b44b-2fc419603c2d AMQPConnectionWorkflow - reporting failure: AMQPConnectionWorkflowFailed: 1 exceptions in all; last exception - AMQPConnectorAMQPHandshakeError: ProbableAuthenticationError: Client was disconnected at a connection stage indicating a probable authentication error: ("ConnectionClosedByBroker: (403) 'ACCESS_REFUSED - Login was refused using authentication mechanism PLAIN. For details see the broker logfile.'",); first exception - None

2023-11-16T11:39:50.841-06:00

Copy
[INFO]	2023-11-16T17:39:50.841Z	8fda8aeb-3351-4569-b44b-2fc419603c2d	Closing transport socket and unlinking: state=3; <ssl.SSLSocket fd=11, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6, laddr=('169.254.76.1', 46510), raddr=('10.191.69.188', 5671)>
[INFO] 2023-11-16T17:39:50.841Z 8fda8aeb-3351-4569-b44b-2fc419603c2d Closing transport socket and unlinking: state=3; <ssl.SSLSocket fd=11, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6, laddr=('169.254.76.1', 46510), raddr=('10.191.69.188', 5671)>

2023-11-16T11:39:50.842-06:00

Copy
[ERROR]	2023-11-16T17:39:50.842Z	8fda8aeb-3351-4569-b44b-2fc419603c2d	Connection workflow failed: AMQPConnectionWorkflowFailed: 1 exceptions in all; last exception - AMQPConnectorAMQPHandshakeError: ProbableAuthenticationError: Client was disconnected at a connection stage indicating a probable authentication error: ("ConnectionClosedByBroker: (403) 'ACCESS_REFUSED - Login was refused using authentication mechanism PLAIN. For details see the broker logfile.'",); first exception - None
[ERROR] 2023-11-16T17:39:50.842Z 8fda8aeb-3351-4569-b44b-2fc419603c2d Connection workflow failed: AMQPConnectionWorkflowFailed: 1 exceptions in all; last exception - AMQPConnectorAMQPHandshakeError: ProbableAuthenticationError: Client was disconnected at a connection stage indicating a probable authentication error: ("ConnectionClosedByBroker: (403) 'ACCESS_REFUSED - Login was refused using authentication mechanism PLAIN. For details see the broker logfile.'",); first exception - None

2023-11-16T11:39:50.859-06:00

Copy
[ERROR]	2023-11-16T17:39:50.842Z	8fda8aeb-3351-4569-b44b-2fc419603c2d	Error in _create_connection().
Traceback (most recent call last):
  File "/opt/python/pika/adapters/blocking_connection.py", line 451, in _create_connection
    raise self._reap_last_connection_workflow_error(error)
pika.exceptions.ProbableAuthenticationError: ConnectionClosedByBroker: (403) 'ACCESS_REFUSED - Login was refused using authentication mechanism PLAIN. For details see the broker logfile.'
[ERROR] 2023-11-16T17:39:50.842Z 8fda8aeb-3351-4569-b44b-2fc419603c2d Error in _create_connection(). Traceback (most recent call last): File "/opt/python/pika/adapters/blocking_connection.py", line 451, in _create_connection raise self._reap_last_connection_workflow_error(error) pika.exceptions.ProbableAuthenticationError: ConnectionClosedByBroker: (403) 'ACCESS_REFUSED - Login was refused using authentication mechanism PLAIN. For details see the broker logfile.'
```
</details>

**Reply by lukebakken** [reply]
> I had printed out the new secret in logs to confirm that it is available to RabbitMQ.

Is RabbitMQ actually USING the new secret? What you have shown so far suggests that RabbitMQ is NOT using the new secret.

I'll ask again - What does RabbitMQ log at the same time your application logs the `ProbableAuthenticationError` error?

**Reply by rshandilya01** [reply]
let me clarify more on 1st question -  
1. RabbitMQ is using Current Secret to get the connection - 
conn = get_mq_connection(current_dict, "set_secret")
2. Once connection is obtained, log shows as below - which tells me that we have an active connection - 
2023-11-16T11:39:50.278-06:00

Copy
[INFO]	2023-11-16T17:39:50.278Z	386b8a39-9b2d-41f2-a5b4-861666db607d	AMQPConnector - reporting success: <SelectConnection OPEN transport=<pika.adapters.utils.io_services_utils._AsyncSSLTransport object at 0x7f634e37a2b0> params=<ConnectionParameters host=b-b3c1ac37-6053-4168-8d04-a3d6d87ccfef.mq.us-east-1.amazonaws.com port=5671 virtual_host=/ ssl=True>>
[INFO] 2023-11-16T17:39:50.278Z 386b8a39-9b2d-41f2-a5b4-861666db607d AMQPConnector - reporting success: <SelectConnection OPEN transport=<pika.adapters.utils.io_services_utils._AsyncSSLTransport object at 0x7f634e37a2b0> params=<ConnectionParameters host=b-b3c1ac37-6053-4168-8d04-a3d6d87ccfef.mq.us-east-1.amazonaws.com port=5671 virtual_host=/ ssl=True>>

2023-11-16T11:39:50.279-06:00

Copy
[INFO]	2023-11-16T17:39:50.279Z	386b8a39-9b2d-41f2-a5b4-861666db607d	AMQPConnectionWorkflow - reporting success: <SelectConnection OPEN transport=<pika.adapters.utils.io_services_utils._AsyncSSLTransport object at 0x7f634e37a2b0> params=<ConnectionParameters host=b-b3c1ac37-6053-4168-8d04-a3d6d87ccfef.mq.us-east-1.amazonaws.com port=5671 virtual_host=/ ssl=True>>
[INFO] 2023-11-16T17:39:50.279Z 386b8a39-9b2d-41f2-a5b4-861666db607d AMQPConnectionWorkflow - reporting success: <SelectConnection OPEN transport=<pika.adapters.utils.io_services_utils._AsyncSSLTransport object at 0x7f634e37a2b0> params=<ConnectionParameters host=b-b3c1ac37-6053-4168-8d04-a3d6d87ccfef.mq.us-east-1.amazonaws.com port=5671 virtual_host=/ ssl=True>>

2023-11-16T11:39:50.279-06:00

Copy
[INFO]	2023-11-16T17:39:50.279Z	386b8a39-9b2d-41f2-a5b4-861666db607d	Connection workflow succeeded: <SelectConnection OPEN transport=<pika.adapters.utils.io_services_utils._AsyncSSLTransport object at 0x7f634e37a2b0> params=<ConnectionParameters host=b-b3c1ac37-6053-4168-8d04-a3d6d87ccfef.mq.us-east-1.amazonaws.com port=5671 virtual_host=/ ssl=True>>
[INFO] 2023-11-16T17:39:50.279Z 386b8a39-9b2d-41f2-a5b4-861666db607d Connection workflow succeeded: <SelectConnection OPEN transport=<pika.adapters.utils.io_services_utils._AsyncSSLTransport object at 0x7f634e37a2b0> params=<ConnectionParameters host=b-b3c1ac37-6053-4168-8d04-a3d6d87ccfef.mq.us-east-1.amazonaws.com port=5671 virtual_host=/ ssl=True>>

**Reply by rshandilya01** [reply]
Let me check the CloudWatch logs to see what RabbitMQ logs at the error time.

**Reply by rshandilya01** [reply]
RabbitMQ log when setting the new Secret - 
2023-11-16T11:39:50.671-06:00

Copy
2023-11-16 17:39:50.283686+00:00 [info] <0.2518.0> connection <0.2518.0> (10.0.0.242:8809 -> 172.18.0.3:5671): user 'admin' authenticated and granted access to vhost '/'
2023-11-16 17:39:50.283686+00:00 [info] <0.2518.0> connection <0.2518.0> (10.0.0.242:8809 -> 172.18.0.3:5671): user 'admin' authenticated and granted access to vhost '/'

2023-11-16T11:39:50.671-06:00

Copy
2023-11-16 17:39:50.301723+00:00 [info] <0.2518.0> connection <0.2518.0> (10.0.0.242:8809 -> 172.18.0.3:5671): user 'admin' updated secret, reason: set_secret AWS Secret Rotation by Lambda
2023-11-16 17:39:50.301723+00:00 [info] <0.2518.0> connection <0.2518.0> (10.0.0.242:8809 -> 172.18.0.3:5671): user 'admin' updated secret, reason: set_secret AWS Secret Rotation by Lambda

2023-11-16T11:39:51.171-06:00

Copy
2023-11-16 17:39:50.323867+00:00 [info] <0.2518.0> closing AMQP connection <0.2518.0> (10.0.0.242:8809 -> 172.18.0.3:5671, vhost: '/', user: 'admin')

**Reply by rshandilya01** [reply]
RabbitMQ log when testing the new Secret - 
2023-11-16T11:39:51.171-06:00

Copy
2023-11-16 17:39:50.841744+00:00 [info] <0.2534.0> accepting AMQP connection <0.2534.0> (10.0.0.242:22801 -> 172.18.0.3:5671)
2023-11-16 17:39:50.841744+00:00 [info] <0.2534.0> accepting AMQP connection <0.2534.0> (10.0.0.242:22801 -> 172.18.0.3:5671)

2023-11-16T11:39:51.171-06:00

Copy
2023-11-16 17:39:50.861505+00:00 [error] <0.2534.0> Error on AMQP connection <0.2534.0> (10.0.0.242:22801 -> 172.18.0.3:5671, state: starting):
2023-11-16 17:39:50.861505+00:00 [error] <0.2534.0> Error on AMQP connection <0.2534.0> (10.0.0.242:22801 -> 172.18.0.3:5671, state: starting):

2023-11-16T11:39:51.171-06:00

Copy
2023-11-16 17:39:50.861505+00:00 [error] <0.2534.0> PLAIN login refused: user 'admin' - invalid credentials
2023-11-16 17:39:50.861505+00:00 [error] <0.2534.0> PLAIN login refused: user 'admin' - invalid credentials

2023-11-16T11:39:55.636-06:00

Copy
2023-11-16 17:39:50.864202+00:00 [info] <0.2534.0> closing AMQP connection <0.2534.0> (10.0.0.242:22801 -> 172.18.0.3:5671)
2023-11-16 17:39:50.864202+00:00 [info] <0.2534.0> closing AMQP connection <0.2534.0> (10.0.0.242:22801 -> 172.18.0.3:5671)

**Reply by rshandilya01** [reply]
I am using the same new Secret when Setting the Secret and Testing the Secret methods. 
conn.update_secret(pending_dict['password'], 'set_secret AWS Secret Rotation by Lambda')
connection = get_mq_connection(pending_dict, 'test_secret')

**Reply by lukebakken** [reply]
When you have your new secret, can you please test opening up a brand-new connection that uses the new secret, instead of using Pika's `update_secret` method?

I would like to _prove_ that the new secret is correct in RabbitMQ. I am asking this because I have no way to reproduce your AWS environment.

In addition, if you have the ability to log into the Management UI using the new secret, or run the `rabbitmqctl authenticate_user admin NEW_SECRET` command, that would be helpful.

**Reply by rshandilya01** [reply]
Luke - here is my set_secret method - it is creating a new connection using current secret and then updating the same with the new secret. Is this what you are asking for? 

I am able to login into the Management UI with the original password that I set up when creating the broker, but not with the new secret, which indicated to me that update secret did not work. 

try:
        # Secrets Manager automatically adds the AWSPREVIOUS staging label to the previous version,
        # so that you retain the last known good version
        previous_dict = get_secret_dict(service_client, arn, "AWSPREVIOUS")
    except (service_client.exceptions.ResourceNotFoundException, KeyError):
        previous_dict = None

    current_dict = get_secret_dict(service_client, arn, "AWSCURRENT")
    pending_dict = get_secret_dict(service_client, arn, "AWSPENDING", token)

    # Now try the current password
    conn = get_mq_connection(current_dict, "set_secret")
    if not conn:
        if previous_dict:
            # If current does not work, try previous.
            conn = get_mq_connection(previous_dict, "set_secret")

    # If we still don't have a connection, raise a ValueError
    if not conn:
        logger.error("set_secret: Unable to log into database with previous or current secret of secret arn %s" % arn)
        raise ValueError("set_secret: Unable to log into database with previous or current secret of secret arn %s" % arn)

    """Update MQ broker with new password"""
    try:
        logger.info("set_secret: Updating Secret in the MQ Broker Connection Adapter")
        conn.update_secret(pending_dict['password'], 'set_secret AWS Secret Rotation by Lambda')
        logger.info("set_secret: Updated Secret in the MQ Broker Connection Adapter")
    except pika.exceptions.ConnectionWrongStateError as connectionerror:
        logger.error(
            "set_secret: ConnectionWrongStateError while updating Secret on RabbitMQ Broker: Connection is not open or available",
            connectionerror)
    except Exception as error:
        logger.error("set_secret: Error when updating Secret on RabbitMQ Broker:", error)
    finally:
        conn.close()
    logger.info("set_secret: Successfully set secret for %s." % arn)

**Reply by lukebakken** [reply]
https://www.rabbitmq.com/oauth2.html#token-expiration

`update_secret` will only update the secret for the _current connection_ (not globally). New connections, or logging into the Management UI, must acquire the new secret before accessing RabbitMQ.

My guess at this point is that you do not have RabbitMQ set up correctly to acquire the new secret.

I doubt I will be able to assist you further unless you can provide an environment that reproduces this issue. `docker compose` is a good option.

@MarcialRosales - is my understanding correct?

**Reply by rshandilya01** [reply]
Not sure what you meant above - use case is pretty simple. Please let me explain again - I am rotating password for the admin user. I am using old password (say, x) to connect to the server with admin user. I am using the same connection to update new password (say, y) and then close the connection. When I am trying to connect to the server again (this time with password as y and user as admin), I am getting the above error. When I try to login to the Management UI with password x and user as admin, I am able to. That indicates to me that update_secret did not work as expected, hence the discussion.

**Reply by lukebakken** [reply, marked as answer]
The `update_secret` method DOES NOT update the password for a user in RabbitMQ.

Please carefully read the docs to which I linked. Updating secrets is generally used in environments that use OAuth2 authentication.

If you are not using OAuth2, your "AWS Lambda rotator" process will have to update the password for the `admin` user (or whatever user you're using) by using either the `rabbitmqctl` command line, or by using the [HTTP API](https://rawcdn.githack.com/rabbitmq/rabbitmq-server/v3.12.8/deps/rabbitmq_management/priv/www/api/index.html).

To update a user's password, you would make a `POST` request to `/api/users/USERNAME` with the correct payload.

**Reply by rshandilya01** [reply]
Thanks Luke - that answers my question. I will use the HTTP API.

**Reply by MarcialRosales** [reply]
That is correct @lukebakken . This method is only meant to "re-authenticate" a live connection, i.e. to change the credentials which have already changed on the backend. This is not meant to change the credentials on the authentication backend. 
If the password has changed in the backend, while the user is connected, the next time the user performs any action over the connection is going to be rejected unless the user has "updated" the password on the live connection.
